man/man1/arr2clm.1
.TH arr2clm 1  "USER COMMANDS"
.SH NAME
arr2clm \- convert 2-D float arrays into CLM file
.SH SYNOPSIS
.B arr2clm
[\-firstyear y] [\-scale factor] [\-id string] [\-float] [\-version v] [\-r value] [\-invalid value] [\-json]
gridfile arrfile clmfile
.SH DESCRIPTION
Program converts binary 2-D float arrays into CLM input data. The array has to be of size 360/lon_resolution x 180/lat_resolution. The latitudes have to start with 90-0.5*lat_resolution, the longitudes with -180+0.5*lon_resolution. The resolution is taken from the grid file.
.SH OPTIONS
.TP
\-firstyear y
Set first year in CLM file, default is 1901.
.TP
\-scale factor
Scaling factor for CLM files. Default is 1.0.
.TP
\-id string
LPJ header string in clm file.
.TP
\-version v
Version of clm header, default is 3.
.TP
\-float
Write float values in clm file, default is short.
.TP
\-invalid value
Value of invalid data in array file, default is -9999.
.TP
\-r value
Value to replace invalid data in CLM file, default is the invalid value.
.TP
\-json
JSON metafile is created with suffix '.json'
.TP
.I gridfile
Filename of grid data file.
.TP
.I arrfile
Filename of array file to convert.
.TP
.I clmfile
Filename of CLM file.
.SH EXAMPLES
.TP
Create CLM data file replacing invalid data by 10.
.B arr2clm
-float -r 10.0 grid.clm array.bin data.clm
.PP
.SH EXIT STATUS
.B arr2clm
returns a zero exit status if the CLM file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
cdf2coord(1), cdf2bin(1), clm2cdf(1), bin2cdf(1), country2cdf(1), regridlpj(1), lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review:

### Added

- Check for valid `swc_bnf` parameters added in `fscanpftpar.c`.
- Wirth et al. (2024)  article added to `REFERENCES`.

### Changed

- `phi_bnf` parameter calculated from `swc_bnf` parameter instead of reading from `pft.cjson` file in `fscanpftpar.c`.

### Removed

- Parameter `phi_bnf` removed from `pft.cjson`.


## [5.10.0] - 2025-08-14

### Contributors

- author: Jens Heinke (heinke@pik-potsdam.de), Sibyll Schaphoff (sibylls@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- option to specify year after which deforestation is no longer feeding product pools but biomass is burnt instead: `luc_burn_startyear` (disabled)
- option to specify costs for N recovery upon turnover (pre-compiler switch `NRECOVERY_COST`, disabled)

### Changed

- parameter changes: 
  - `alphaa` from 0.5 to 0.7 for natural vegetation; 
  - `theta` from 0.9 to 0.95; 
  - `fn_turnover` PFT1: 0.5, PFT2: 0.5, PFT3: 0.7, PFT4: 0.3, PFT5: 0.3, PFT6: 0.7, PFT7: 0.3, PFT8: 0.3; updated values from [Sophia et al. 2024](https://doi.org/10.5194/bg-21-4169-2024) for all other PFTs except crops
  - `cnratio_leaf.low/median/high` for all PFTs updated from TRY data
  - `b` (ratio of dark respiration Rd to maximum photosynthetic capacity Vcmax) set to 0.031 for C3 and C4 plants (value for 25°C, from [Wang et al. 2020](https://doi.org/10.1111/gcb.14980), consistent with TRY data)
  - `aphen_min` from 60 to 90 and `aphen_max` from 245 to 180
  - `nfixpot` from 0.01 to 0.5 (middle of proposed range in [Yu and Zhuang 2020](https://doi.org/10.5194/bg-17-3643-2020))
  - `turnover` parameter for leaves&roots set to 2 for TeNE and 3 BoNE 
  - temperature boundary between temperate and boreal zone changed from -2°C to -5°C
- acclimation of `b` to mean vegetation period temperature (following [Wang et al. 2020](https://doi.org/10.1111/gcb.14980))
- calculation of structural leaf N content (leaf N independent of Vcmax) based on `ncleaf.low` in both `ndemand_xx` and `vmaxlimit_xx` (was `ncleaf.medium` in `ndemand_xx`)
- scaling of vmax with f_LAI removed
- N recovery from turnover directly added to `bm_inc.nitrogen`
- N uptake now separated for uptake of NH4 and NO3 with parameters from  [Craig et al. 2025](https://doi.org/10.1111/nph.70140) (median of values)
- trees are burnt (instead of added to the litter) when running with `luc_timber=FALSE`
- beginning of vegetative period for tropical raingreen trees set at beginning of 6 month period with highest precipitation sum
- leaf shedding for deciduous trees induced after `aphen_max` or when `phen<0.1` after `aphen_min`
- `phen` permanently set to 1 for tropical evergreen trees

### Removed

### Fixed

- corrected conversion of vmax in computation of N demand, which was incorrectly scaled with `24/daylength`
- corrected temperature sensitivity of N demand (`k_temp` changed from 0.02 to 0.0693)
- corrected `f_NCplant` to comply with equation from [Smith et al. 2014](https://doi.org/10.5194/bg-11-2027-2014)
- corrected phenology of trees:
  - leaf turnover of raingreen trees at leaf shedding (same as for summergreen trees, had continuous daily leaf turnover like evergreen trees)
  - `phen` set to zero at leaf shedding until restart of phenological cycle to prevent phen>1 without leaf carbon 


## [5.9.28] - 2025-07-31

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Maik Billing (billing@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de), Sebastian Ostberg(ostberg@pik-potsdam.de), Jens Heinke (heinke@pik-potsdam.de)

### Added

- Option `-h` added to utilities `addheader`, `cdf2bin`, `regridclm`, `regridsoil` and `printclm` to print help text.
- Notice added in `configure.sh` that `make clean; make all` has to be performed after change in configuration.
- Support for short datatype added in `cmpbin` utility.
- Shell script `allbin2cdf` added to convert all binary raw output files in a directory to NetCDF files.
- Option `-nounit` added to `bin2cdf` to set unit if unit is defined in the metafile as an empty string.

### Changed

- If option `-ncell 0` is used in `addheader` utility then the number of cells is calculated from the file size of the binary file.
- Utilities `grid2clm` and `cft2clm` are replaced by alias to `addheader`.
- File `default.md` converted from DOS to Linux format.
- If scaling factor is set as an option in `bin2cdf` then this scaling factor is used insteads that one defined in the JSON metafile.
- Man pages updated.

### Fixed

- Missing `free(axis->comment)` added in `freeaxis()` to avoid memory leak.
- Missing deallocation of list added in `open_config.c`.
- Function `nc_close()` replaced by `closeclimate_netcdf()` in `cdf2clm.c` and `cdf2bin.c` to avoid memory leak.
- Argument for error message corrected in `getcellindex.c` and `getcountry.c`.
- Typo in man page of `configure.sh` fixed.
- Comment for `pft` initialized in `initsetting_netcdf.c`.
- Access to undefined `pft_name.comment` removed in `bin2cdf.c` to avoid SEGV.

src/utils/arr2clm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                     a  r  r  2  c  l  m  .  c                                  \n**/
/**                                                                                \n**/
/**     Program converts 2-D float arrays into CLM format                          \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/
#include "lpj.h"

#define USAGE "Usage: arr2clm [-firstyear y] [-version v] [-float] [-scale s] [-id s] [-invalid i] [-r v] gridfile arrfile clmfile\n"

int main(int argc,char **argv)
{
  int iarg,i,version,count,nlon,nlat;
  Bool isfloat;
  Coordfile coordfile;
  Coord *coords;
  Coord resolution;
  Filename coord_filename;
  Header header;
  FILE *in,*out;
  float *data,value,invalid,replace;
  short svalue;
  char *id,*endptr;
  isfloat=FALSE;
  header.scalar=1;
  header.firstyear=1901;
  header.firstcell=0;
  header.nbands=1;
  header.nstep=1;
  header.timestep=1;
  header.order=CELLYEAR;
  id=LPJ_CLIMATE_HEADER;
  version=LPJ_CLIMATE_VERSION;
  invalid=replace=-9999;
  for(iarg=1;iarg<argc;iarg++)
  {
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-float"))
        isfloat=TRUE;
      else if(!strcmp(argv[iarg],"-scale"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Error: Missing argument after option '-scale'.\n"
                  USAGE);
          return EXIT_FAILURE;
        }
        header.scalar=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Error: Invalid number '%s' for option '-scale'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
        if(header.scalar==0)
        {
          fputs("Scale factor must not be zero.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-r"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Error: Missing argument after option '-r'.\n"
                  USAGE);
          return EXIT_FAILURE;
        }
        replace=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Error: Invalid number '%s' for option '-r'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-invalid"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Error: Missing argument after option '-invalid'.\n"
                  USAGE);
          return EXIT_FAILURE;
        }
        replace=invalid=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Error: Invalid number '%s' for option '-invalid'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-firstyear"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Error: Missing argument after option '-firstyear'.\n"
                  USAGE);
          return EXIT_FAILURE;
        }
        header.firstyear=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Error: Invalid number '%s' for option '-firstyear'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-version"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Error: Missing argument after option '-version'.\n"
                  USAGE);
          return EXIT_FAILURE;
        }
        version=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Error: Invalid number '%s' for option '-version'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-id"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Missing argument after option '-id'.\n"
                 USAGE);
          return EXIT_FAILURE;
        }
        id=argv[++iarg];
      }
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n",argv[iarg]);
        fputs(USAGE,stderr);
      }
    }
    else
     break;
  }
  if(argc<iarg+3)
  {
    fprintf(stderr,"Missing arguments.\n"
            USAGE);
    return EXIT_FAILURE;
  }
  coord_filename.name=argv[iarg];
  coord_filename.fmt=CLM;
  coordfile=opencoord(&coord_filename,TRUE);
  if(coordfile==NULL)
    return EXIT_FAILURE;
  coords=newvec(Coord,numcoord(coordfile));
  if(coords==NULL)
  {
    printallocerr("coords");
    return EXIT_FAILURE;
  }
  getcellsizecoord(&header.cellsize_lon,&header.cellsize_lat,coordfile);
  resolution.lat=header.cellsize_lat;
  resolution.lon=header.cellsize_lon;
  nlon=(int)(360/resolution.lon);
  nlat=(int)(180/resolution.lat);
  header.ncell=numcoord(coordfile);
  for(i=0;i<numcoord(coordfile);i++)
    readcoord(coordfile,coords+i,&resolution);
  closecoord(coordfile);

  in=fopen(argv[iarg+1],"rb");
  if(in==NULL)
  {
    free(coords);
    printfopenerr(argv[iarg+1]);
    return EXIT_FAILURE;
  }
  header.datatype=(isfloat) ? LPJ_FLOAT : LPJ_SHORT;
  header.nyear=0;
  out=fopen(argv[iarg+2],"wb");
  if(out==NULL)
  {
    free(coords);
    printfcreateerr(argv[iarg+2]);
    return EXIT_FAILURE;
  }
  fwriteheader(out,&header,id,version);
  data=newvec(float,nlon*nlat);
  if(data==NULL)
  {
    free(coords);
    printallocerr("data");
    return EXIT_FAILURE;
  }
  if(getfilesizep(in) % (sizeof(float)*nlat*nlon))
    fprintf(stderr,"Warning: File size of '%s' is not multiple of %d x %d.\n",
            argv[iarg+1],nlon,nlat);
  while(fread(data,sizeof(float),nlat*nlon,in)==nlat*nlon)
  {
    count=0;
    for(i=0;i<header.ncell;i++)
    {
      value=data[nlon*(int)((90-resolution.lat*0.5-coords[i].lat)/resolution.lat)+(int)((coords[i].lon+180-resolution.lon*0.5)/resolution.lon)];
      if(value==invalid)
      {
        value=replace;
        count++;
      }
      if(isfloat)
        fwrite(&value,sizeof(float),1,out);
      else
      {
        if(round(value/header.scalar)<SHRT_MIN || round(value/header.scalar)>SHRT_MAX)
          fprintf(stderr,"Warning: Data overflow %g for cell %d.\n ",value/header.scalar,i);
        svalue=(short)round(value/header.scalar);
        fwrite(&svalue,sizeof(short),1,out);
      }
    }
    if(count)
      fprintf(stderr,"Warning: %d invalid cells in year %d, set to %g.\n",
              count,header.firstyear+header.nyear,replace);
    header.nyear++;
  }
  free(data);
  free(coords);
  printf("Number of years: %d\n",header.nyear);
  rewind(out);
  fwriteheader(out,&header,id,version);
  fclose(out);
  fclose(in);
  return EXIT_SUCCESS;
} /* of 'main' */
